### PR TITLE
Add shopingcart model implementation

### PR DESCRIPTION
**Pull request by ValentynSharshon**

### CODE REVIEW

**Review by fedaktaras — APPROVED** [review]
Looks really good.
You have some good features I would like to add to mine project.

File: src/main/java/com/gmail/woosay333/onlinebookstore/controller/ShoppingCartController.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        return shoppingCartService.addToCart(request);
+    }
+
+    @PreAuthorize("hasRole('USER')")
+    @PutMapping("/cart-items/{cartItemId}")
+    @ResponseStatus(HttpStatus.OK)
+    @Operation(summary = "Update a cart item.",
+            description = "Update an existing cart item in the user's shopping cart by its ID.")
+    public CartItemDto updateCartItem(@PathVariable Long cartItemId,
+                                      @RequestBody @Valid CartItemUpdateRequestDto request) {
```
Comment (andrii-hoienko) [inline src/main/java/com/gmail/woosay333/onlinebookstore/controller/ShoppingCartController.java]: Inconsistency in names, you have CreateCartItemRequestDto and CartItemUpdateRequestDto, please decide where to put word that belong to operation. CreateCartItemRequestDto and UpdateCartItemRequestDto or CartItemUpdateRequestDto and CartItemCreateRequestDto

File: src/main/java/com/gmail/woosay333/onlinebookstore/entity/ShoppingCart.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.SQLRestriction;
+
+@Entity
+@Table(name = "shopping_carts")
+@Data
+@NoArgsConstructor
+@SQLDelete(sql = "UPDATE shopping_carts SET is_deleted = true WHERE id = ?")
+@SQLRestriction(value = "is_deleted = false")
+
```
Comment (andrii-hoienko) [inline src/main/java/com/gmail/woosay333/onlinebookstore/entity/ShoppingCart.java]:
```suggestion
```

File: src/main/java/com/gmail/woosay333/onlinebookstore/service/shopping/cart/impl/ShoppingCartServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+                shoppingCart.getId())
+                .orElseThrow(() -> new EntityNotFoundException(
+                        String.format("Can't find cart item with id: %d", cartItemId)));
+        cartItem.setQuantity(requestDto.getQuantity());
+        return cartItemMapper.toDto(cartItemRepository.save(cartItem));
+    }
+
+    @Override
+    public void deleteCartItem(Long cartItemId) {
+        cartItemRepository.deleteById(cartItemId);
```
Comment (andrii-hoienko) [inline src/main/java/com/gmail/woosay333/onlinebookstore/service/shopping/cart/impl/ShoppingCartServiceImpl.java]: Please check is exists before delete

File: src/main/java/com/gmail/woosay333/onlinebookstore/service/shopping/cart/impl/ShoppingCartServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    @Override
+    @Transactional
+    public ShoppingCartDto addToCart(CreateCartItemRequestDto cartItemRequestDto) {
+        User user = userService.getAuthenticatedUser();
+        Book book = bookRepository.findById(cartItemRequestDto.getBookId())
+                .orElseThrow(() -> new EntityNotFoundException(
+                        String.format("Book with id: %d not found",
+                                cartItemRequestDto.getBookId())));
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(user.getId())
+                .orElseGet(() -> createNewShoppingCart(user));
```
Comment (andrii-hoienko) [inline src/main/java/com/gmail/woosay333/onlinebookstore/service/shopping/cart/impl/ShoppingCartServiceImpl.java]: Code duplication

File: src/main/java/com/gmail/woosay333/onlinebookstore/service/shopping/cart/impl/ShoppingCartServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+        CartItem cartItem = createCartItem(book, shoppingCart, cartItemRequestDto.getQuantity());
+        cartItemRepository.save(cartItem);
+        shoppingCart.getCartItems().add(cartItem);
+        return shoppingCartMapper.toDto(shoppingCart);
```
Comment (andrii-hoienko) [inline src/main/java/com/gmail/woosay333/onlinebookstore/service/shopping/cart/impl/ShoppingCartServiceImpl.java]: I don`t see the logic that saves shopping cart after it was updated with new cartItem

File: src/main/java/com/gmail/woosay333/onlinebookstore/service/shopping/cart/impl/ShoppingCartServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    public CartItemDto updateCartItem(Long cartItemId, CartItemUpdateRequestDto requestDto) {
+        ShoppingCart shoppingCart = shoppingCartRepository
+                .findByUserId(userService.getAuthenticatedUser().getId())
+                .orElseThrow(() -> new EntityNotFoundException(
+                        String.format("There is no cart belonging to user with id: %d",
+                                userService.getAuthenticatedUser().getId())));
+        CartItem cartItem = cartItemRepository.findByIdAndShoppingCartId(cartItemId,
+                shoppingCart.getId())
+                .orElseThrow(() -> new EntityNotFoundException(
+                        String.format("Can't find cart item with id: %d", cartItemId)));
```
Comment (andrii-hoienko) [inline src/main/java/com/gmail/woosay333/onlinebookstore/service/shopping/cart/impl/ShoppingCartServiceImpl.java]: This could be done using one db request

**Review by andrii-hoienko — CHANGES_REQUESTED** [review]
Good job! Left some comments

File: src/main/java/com/gmail/woosay333/onlinebookstore/service/shopping/cart/impl/ShoppingCartServiceImpl.java (hunk 5 above)
Comment (ValentynSharshon) [reply src/main/java/com/gmail/woosay333/onlinebookstore/service/shopping/cart/impl/ShoppingCartServiceImpl.java]: According to the logic of my application, the creation of a shopping cart occurs at the moment the user adds an item to the basket. Did I understand you correctly that I need to move the logic of creating a shopping cart to the moment the user registers in the system?

**Pull request merged**